### PR TITLE
nnpdf: init at 4.0.4

### PR DESCRIPTION
**Pull request by veprbl**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/validphys2/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+buildPythonPackage rec {
+  pname = "validphys2";
+  version = "4.0";
+  format = "setuptools";
+
+  inherit (nnpdf) src;
+
+  prePatch = ''
+    cd validphys2
```
Comment (dotlambda) [inline pkgs/development/python-modules/validphys2/default.nix:15]:
```suggestion
  sourceRoot = "${src.name}/validphys2";
```

File: pkgs/development/python-modules/validphys2/default.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  '';
+
+  postPatch = ''
+    substituteInPlace src/validphys/version.py \
+      --replace '= __give_git()' '= "${version}"'
+  '';
+
+  propagatedBuildInputs = [
+    nnpdf
```
Comment (dotlambda) [inline pkgs/development/python-modules/validphys2/default.nix:24]: Only Python packages should be in `propagatedBuildInputs`.

**Review by dotlambda — COMMENTED** [review]
Validphys should specify `checkInputs` or `doCheck` or `checkPhase` (inclusive or).

File: pkgs/development/python-modules/validphys2/default.nix (hunk 1 above)
Comment (veprbl) [reply pkgs/development/python-modules/validphys2/default.nix:15]: `sourceRoot` is broken by design. Your varian will break when overriding with a local path like `src = lib.cleanSources ./nnpdf` and possibly also for overriding to `fetchurl` and `fetchgit`.

File: pkgs/development/python-modules/validphys2/default.nix (hunk 2 above)
Comment (veprbl) [reply pkgs/development/python-modules/validphys2/default.nix:24]: This now is a reference to `pythonPackages.nnpdf`

File: pkgs/development/python-modules/validphys2/default.nix (hunk 1 above)
Comment (dotlambda) [reply pkgs/development/python-modules/validphys2/default.nix:15]: You should at least use `postUnpack`.

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -5719,6 +5719,10 @@ in {
 
   nltk = callPackage ../development/python-modules/nltk { };
 
+  nnpdf = toPythonModule (pkgs.nnpdf.override {
+    python3 = python;
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix:5723]: If I remember correctly only
```suggestion
    python3 = self.python;
```
works.

Comment (veprbl) [reply pkgs/top-level/python-packages.nix:5723]: I doubt it makes any difference
```
# ag python3 pkgs/top-level/python-packages.nix 
42:    inherit namePrefix;     # We want Python libraries to be named like e.g. "python3.6-${name}"
552:  antlr4-python3-runtime = callPackage ../development/python-modules/antlr4-python3-runtime {
584:  appleseed = toPythonModule (pkgs.appleseed.override { python3 = self.python; });
833:  avro-python3 = callPackage ../development/python-modules/avro-python3 { };
3317:    python3 = python;
4626:    python3 = python;
4767:    python3 = python;
4790:          outputsToInstall = [ "py" ]; # The package always builds python3 bindings
4794:    (p: p.override { python3 = python; })
4851:        python3 = python;
4929:    python3 = python;
5205:  mecab-python3 = callPackage ../development/python-modules/mecab-python3 { };
5226:    python3Packages = self;
5245:  meson = toPythonModule ((pkgs.meson.override { python3 = python; }).overrideAttrs
5293:    python3 = python;
5335:    python3 = python;
5669:    python3 = python;
5674:    python3Packages = self;
5723:    python3 = python;
7884:    python3 = self.python;
8277:  python3-application = callPackage ../development/python-modules/python3-application { };
8279:  python3-eventlib = callPackage ../development/python-modules/python3-eventlib { };
8281:  python3-gnutls = callPackage ../development/python-modules/python3-gnutls { };
8283:  python3-openid = callPackage ../development/python-modules/python3-openid { };
8287:  python3-saml = callPackage ../development/python-modules/python3-saml { };
8375:    meson = pkgs.meson.override { python3 = self.python; };
8944:    python3Packages = self;
9096:    python3 = python;
9356:    python3 = self.python;
11312:    python3 = python;
11318:    python3 = python;
```

File: pkgs/development/python-modules/validphys2/default.nix (hunk 1 above)
Comment (veprbl) [reply pkgs/development/python-modules/validphys2/default.nix:15]: `postUnpack` is called before `cd $sourceRoot`
https://github.com/NixOS/nixpkgs/blob/ea69f79829972187ecf6c1eaede4ce7a15d81736/pkgs/stdenv/generic/setup.sh#L1364-L1374

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 173156` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>nnpdf (python39Packages.nnpdf)</li>
    <li>python310Packages.nnpdf</li>
    <li>python310Packages.validphys2</li>
    <li>validphys2 (python39Packages.validphys2)</li>
  </ul>
</details>

**Comment by veprbl** [comment]
Thanks!

**Pull request merged**